src/paint/paint.c

```diff
@@ -59,17 +59,153 @@ bool sub_98196C(
+			rotate_map_coordinates(&boundBox.x, &boundBox.y, MAP_ELEMENT_DIRECTION_WEST);
+			break;
+
+		case 1:
+			rotate_map_coordinates(&coord_3d.x, &coord_3d.y, MAP_ELEMENT_DIRECTION_SOUTH);
+
+			boundBox.x--;
+			rotate_map_coordinates(&boundBox.x, &boundBox.y, MAP_ELEMENT_DIRECTION_SOUTH);
+			break;
+
+		case 2:
+			rotate_map_coordinates(&coord_3d.x, &coord_3d.y, MAP_ELEMENT_DIRECTION_EAST);
+			rotate_map_coordinates(&boundBox.x, &boundBox.y, MAP_ELEMENT_DIRECTION_EAST);
+			break;
+
+		case 3:
+			rotate_map_coordinates(&coord_3d.x, &coord_3d.y, MAP_ELEMENT_DIRECTION_NORTH);
+
+			boundBox.y--;
+			rotate_map_coordinates(&boundBox.x, &boundBox.y, MAP_ELEMENT_DIRECTION_NORTH);
+			break;
+	}
+
+	coord_3d.x += RCT2_GLOBAL(0x9DE568, sint16);
+	coord_3d.y += RCT2_GLOBAL(0x9DE56C, sint16);
+
+	ps->bound_box_x_end = coord_3d.x + boundBox.x;
+	ps->bound_box_y_end = coord_3d.y + boundBox.y;
+
+	// TODO: check whether this is right. edx is ((bound_box_length_z + z_offset) << 16 | z_offset)
+	ps->bound_box_z = coord_3d.z;
+	ps->bound_box_z_end = (boundBox.z + coord_3d.z);
+
+	rct_xy16 map = coordinate_3d_to_2d(&coord_3d, rotation);
+
+	ps->x = map.x;
+	ps->y = map.y;
+
+	sint16 left = map.x + g1Element->x_offset;
+	sint16 bottom = map.y + g1Element->y_offset;
+
+	sint16 right = left + g1Element->width;
+	sint16 top = bottom + g1Element->height;
+
+	RCT2_GLOBAL(0xF1AD1C, sint16) = left;
+	RCT2_GLOBAL(0xF1AD1E, sint16) = bottom;
+
+	rct_drawpixelinfo *dpi = RCT2_GLOBAL(0x140E9A8, rct_drawpixelinfo*);
+
+	if (right <= dpi->x) return false;
+	if (top <= dpi->y) return false;
+	if (left >= (dpi->x + dpi->width)) return false;
+	if (bottom >= (dpi->y + dpi->height)) return false;
+
+	ps->flags = 0;
+	ps->bound_box_x = coord_3d.x;
+	ps->bound_box_y = coord_3d.y;
+	ps->attached_ps = NULL;
+	ps->var_20 = NULL;
+	ps->sprite_type = RCT2_GLOBAL(RCT2_ADDRESS_PAINT_SETUP_CURRENT_TYPE, uint8);
+	ps->var_29 = RCT2_GLOBAL(0x9DE571, uint8);
+	ps->map_x = RCT2_GLOBAL(0x9DE574, uint16);
+	ps->map_y = RCT2_GLOBAL(0x9DE576, uint16);
+	ps->mapElement = RCT2_GLOBAL(0x9DE578, rct_map_element*);
+
+	RCT2_GLOBAL(0xF1AD28, paint_struct*) = ps;
+
+	sint32 edi;
+	switch (rotation) {
+		case 0:
+			edi = coord_3d.y + coord_3d.x;
+			break;
+
+		case 1:
+			edi = coord_3d.y - coord_3d.x + 0x2000;
+			break;
+
+		case 2:
+			edi = -(coord_3d.y + coord_3d.x) + 0x4000;
+			break;
+
+		case 3:
+			edi = coord_3d.x - coord_3d.y + 0x2000;
+			break;
+	}
+
+	if (edi < 0) {
+		edi = 0;
+	}
+
+	edi /= 32;
+	edi = min(edi, 0x1FF); // 511
+
+	ps->var_18 = edi;
+
+	paint_struct *old_ps = RCT2_ADDRESS(0x00F1A50C, paint_struct*)[edi];
+	RCT2_ADDRESS(0x00F1A50C, paint_struct*)[edi] = ps;
+	ps->next_quadrant_ps = old_ps;
+
+	if ((uint16)edi < RCT2_GLOBAL(0x00F1AD0C, uint32)) {
+		RCT2_GLOBAL(0x00F1AD0C, uint32) = edi;
+	}
+
+	if ((uint16)edi > RCT2_GLOBAL(0x00F1AD10, uint32)) {
+		RCT2_GLOBAL(0x00F1AD10, uint32) = edi;
+	}
+
+	RCT2_GLOBAL(0xEE7888, uint32) += 0x34;
+
+	return true;
```

Comment:
Can you change all of these to return the `paint_struct*` as already done in your path PR.